### ISSUE TITLE
[Bug]: training.warmup_steps config key is dead - no LR scheduler implemented

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#51]
## Problem Statement

`configs/default.yaml` defines `training.warmup_steps: 100` and the config schema includes it, but no code in the training pipeline consumes this value. The `Trainer` creates a bare `AdamW` optimizer with no learning rate scheduler attached.

Grepping `nightmarenet/training/` for `warmup`, `lr_scheduler`, `LambdaLR`, `CosineAnnealing`, or `get_linear_schedule_with_warmup` returns zero matches.

This means the model jumps to full learning rate from step 0. For fine-tuning pretrained transformers (the primary use case), this causes gradient instability in early steps, especially in the nightmare phase which already applies a 2x LR multiplier.

**Relevant files:**
- `nightmarenet/training/trainer.py` line 149 (optimizer creation, no scheduler)
- `configs/default.yaml` line 28 (`warmup_steps: 100` - dead config)

## Proposed Solution

Add a learning rate warmup scheduler:

1. Use `transformers.get_linear_schedule_with_warmup` (already a dependency) or `torch.optim.lr_scheduler.LambdaLR` with a linear warmup lambda
2. Compute `total_steps` from `num_cycles * phases * epochs * batches_per_epoch`
3. Step the scheduler after each optimizer step in all phase runners
4. Config: `training.warmup_steps` (already exists), optionally add `training.lr_schedule: "linear_warmup"` for future extensibility

**Files affected:**
- `nightmarenet/training/trainer.py` - create scheduler, pass to phases
- `nightmarenet/training/phases.py` - step scheduler in WakePhase, DreamPhase, NightmarePhase training loops

**Dependencies:** None new (`transformers` already installed).

## Acceptance Criteria

- [ ] `training.warmup_steps` config value is consumed and produces a warmup ramp
- [ ] LR starts at 0 and reaches `learning_rate` after `warmup_steps` steps
- [ ] LR schedule persists across phases within a cycle (doesn't reset per phase)
- [ ] Nightmare phase LR multiplier is applied on top of the scheduled LR (not the base LR)
- [ ] Test: verify LR at step 0 < LR at step warmup_steps
- [ ] No behavior change when `warmup_steps: 0`

## Scope / Difficulty Estimate

Small-medium. ~2 hours. Standard transformers scheduler integration.

**Comment by Ruchira807** [comment]
@Adit-Jain-srm I would like to contribute to this issue under ECSOC'26

**Comment by RitikaSen27** [comment]
Hi @Adit-Jain-srm could you please assign this to me?

**Comment by Adit-Jain-srm** [comment]
Assigned, @Ruchira807 
Thank you @RitikaSen27 for showing interest, please request assignment on other issues.

**Comment by Ruchira807** [comment]
@Adit-Jain-srm While looking through the current codebase, I noticed that main now includes a phase scheduler (create_scheduler_from_config) that wasn't mentioned in the issue, although there is still no optimizer LR warmup scheduler.

Before implementing this, could you clarify the preferred approach?

Should the LR warmup scheduler be:

created separately (e.g. transformers.get_linear_schedule_with_warmup) alongside the existing phase scheduler, or
integrated into the current scheduling system?

Also, since the dataloaders are only available in Trainer.train(), would you prefer computing num_training_steps there instead of during Trainer.__init__?

I'd like to make sure the implementation matches the current architecture.

**Comment by Adit-Jain-srm** [comment]
Good research, @Ruchira807. You're right that `create_scheduler_from_config` is a **phase scheduler** (controls the wake/dream/nightmare/compress cycle order) - it has nothing to do with learning rate scheduling. The `warmup_steps` config key remains dead.

To answer your questions:

1. **Create the LR scheduler separately.** Use `torch.optim.lr_scheduler.LambdaLR` with a linear warmup lambda (avoids adding a `transformers` scheduler dependency for something this simple). Keep it independent from the phase scheduler - they operate at different levels (phase scheduler = which phase runs, LR scheduler = what the learning rate is at each optimizer step).

2. **Compute `num_training_steps` in `Trainer.train()`**, not `__init__`. You're correct that dataloaders aren't available until then. Calculate it as:
   ```
   steps_per_epoch = len(dataloader) // gradient_accumulation_steps
   total_steps = sum(epochs_per_phase) * steps_per_epoch
   ```
   Or simpler: just use a warmup-only schedule (linear from 0 to `learning_rate` over `warmup_steps`, constant after) which doesn't need `total_steps` at all.

3. **Pass the scheduler to each phase runner** and call `scheduler.step()` after every `optimizer.step()`. The scheduler should persist across phases within a cycle (don't recreate it per phase).

4. **Nightmare LR multiplier:** Apply it by adjusting the optimizer param group's base LR before the nightmare phase starts, so the warmup schedule scales proportionally. Don't multiply the scheduler's output post-hoc.

Architecture summary:
```
Trainer.__init__:  optimizer = AdamW(...)
Trainer.train():   lr_scheduler = LambdaLR(optimizer, lr_lambda=warmup_lambda)
                   pass lr_scheduler to each phase runner
Phase.run():       after optimizer.step() -> lr_scheduler.step()
```

Go ahead with this approach.
